DM-40303: Fix pydantic v2 warnings

timj opened the pull request:
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Replace some comprehensions with simpler constructors

timj committed:
Simplify some list/tuple constructors

timj committed:
Simplify dict constructor

timj committed:
Use model_validate rather than parse_raw for Pydantic

timj committed:
Target 3.11 for pre-commit and black

timj committed:
Change pydantic.dict() -> pydantic.model_dump()

timj committed:
Migrate pydantic.parse_raw() to model_validate_json()

timj committed:
Use pydantic.model_dump_json() instead of .json()

timj committed:
Use v2 model validator for DatasetRef

Now use a single model validator that checks everything because
we are checking multiple keys.

timj committed:
Fix two more @validators

timj committed:
Use examples as list in Field schema, not example as string

timj committed:
Migrate schema_extra to json_schema_extra for pydantic v2

timj committed:
Replace allow_mutation with frozen in pydantic v2

timj committed:
Allow v1 models to still use .json() for serialization

Only v2 models and v1 models with the compatibility class
support model_dump_json().

andy-slac commented on python/lsst/daf/butler/core/datasets/ref.py:
```diff
@@ -179,24 +180,35 @@ class SerializedDatasetRef(_BaseModelCompat):
+        # seems to trigger with the datasetType field not yet set.
+        @pydantic.model_validator(mode="before")  # type: ignore[attr-defined]
+        @classmethod
+        def check_consistent_parameters(cls, data: Any) -> Any:
```
Should it take `data: dict[str, Any]` and return the same type?

timj replied:
That is probably okay. I was following the pydantic porting guide.

andy-slac commented on python/lsst/daf/butler/datastores/chainedDatastore.py:
```diff
@@ -1064,7 +1064,7 @@ def transfer_from(
             # so we have to filter again based on what the source datastore
             # understands.
-            known_to_source = source_child.knows_these([ref for ref in refs])
+            known_to_source = source_child.knows_these(list(refs))
```
Not sure you need `list` here, `knows_these` says it takes `Iterable`, and `refs` is Iterable here too. I think this code already assumes that you can iterate many times over `refs`, which is probably not true for `Iterable` in general.

timj replied:
I think I was forcing it to a list originally so that we could go through it multiple times, which probably means the annotation is wrong. Should it be Sequence?

andy-slac replied:
We can keep `Iterable` but then make sure that code can handle single-pass iterables. This code in chainedDatastore cannot, it calls `set(refs)`  few lines above. fileDatastore also doing similar thing in `knows_these`. Or alternatively we should use Sequence everywhere.

andy-slac commented on python/lsst/daf/butler/formatters/file.py:
```diff
@@ -175,9 +175,9 @@ def _coerceBuiltinType(self, inMemoryDataset: Any, writeStorageClass: StorageCla
         ):
             # Try different ways of converting to the required type.
-            if hasattr(writeStorageClass.pytype, "parse_obj"):
+            if hasattr(writeStorageClass.pytype, "model_validate"):
```
Hmm, would it be safer to check `isinstance` instead with pydantic base class? `model_validate` is probably not super-popular method name but even that could potentially clash in some random class?

timj replied:
I see I'm already using isinstance later so it probably does no harm to say this is explicitly the pydantic path. The best bit about this comment here is that it explains to me why ci_hsc_gen3 failed. I was trying to understand why ScarletModelData was not working properly and it turns out it's because that class tried to emulate pydantic by having a `parse_obj` and `json()` method and so `isinstance` here would break it. It does though tell me how to fix the scarlet problem, or at least I have to decide whether to keep parse_obj here or change `ScarletModelData` to use `model_validate` and `model_dump_json`...

andy-slac reviewed (approved):
Looks good, a couple of minor comments.

timj committed:
Change file formatter to support v1 and v2 pydantic constructors

The pydantic models that use the v1 compatibility class work
fine with model_validate but there are some v1 models that do
not inherit from that and also some classes like ScarletModelData
that emulate the pydantic v1 methods but are not using pydantic.

Therefore we need to still check parse_obj.

timj committed:
Add news fragment
